### ISSUE TITLE
Keep repeating a test until all tests are finished

### ISSUE THREAD

**Issue by YeskaNova** [nunit/nunit#3814]
Hi,

I have some tests that run in parallel to each other, I need some/all of them to keep repeating themselves until all the tests are run at least 1 time ( or n times ).

Is it possible to achieve that with the current Api?

Yeska

**Comment by Dreamescaper** [comment]
Could you explain your use case? Why do you need that?

**Comment by YeskaNova** [comment]
@Dreamescaper My tests are run by my task processor, each test is independent and tests one feature of the app. Still, in the real scenario, the task processor will run tasks in parallel as they are called by end-users and those tasks will use many resources at the same time. I need to stress-test the task processor to be sure that any two features can run in parallel. This thread-safety is well tested with some unit tests that only test the isolation of each task, but I don't want to rely only on unit tests for this matter as tasks can break this isolation when they access shared resources like IO/Db, so I want to stress test it too. To be sure that any task can be run in parallel with another task, I want all the tests to keep repeating themselves until all tests are run at least one time. It's not my first line of testing, it's rather reusing the already defined tests to stress test the engine.

**Comment by Dreamescaper** [comment]
I don't think that it's possible with current API.
You can probably create a separate console project, which would run tests in a loop (e.g. via [NUnitLite Runner](https://docs.nunit.org/articles/nunit/running-tests/NUnitLite-Runner.html)), and run this application in parallel with running your actual tests.

**Comment by YeskaNova** [comment]
@Dreamescaper Well, I managed to make it work I think, but I didn't test it heavly.

I took the code of RepeatAttribute, and changed this 

https://github.com/nunit/nunit/blob/0296dad2f7a28928b9c4721402ac8747f59afac9/src/NUnitFramework/framework/Attributes/RepeatAttribute.cs#L75

with

```C#
            private static IDictionary<IWorkItemDispatcher, int> _dispatchers = new Dictionary<IWorkItemDispatcher, int>();
            /// <summary>
            /// Runs the test, saving a TestResult in the supplied TestExecutionContext.
            /// </summary>
            /// <param name="context">The context in which the test should run.</param>
            /// <returns>A TestResult</returns>
            public override TestResult Execute(TestExecutionContext context)
            {
                lock (_dispatchers)
                    if (_dispatchers.TryGetValue(context.Dispatcher, out var val))
                        _dispatchers[context.Dispatcher] = val + repeatCount;
                    else
                        _dispatchers[context.Dispatcher] = repeatCount;

                int count = repeatCount;

                while (true)
                {
                    context.CurrentResult = innerCommand.Execute(context);

                    if (--count >= 0)
                    {
                        lock (_dispatchers)
                            --_dispatchers[context.Dispatcher];
                    }

                    lock (_dispatchers)
                        if (_dispatchers[context.Dispatcher] == 0)
                            break;

                    if (context.CurrentResult.ResultState != ResultState.Success)
                    {
                        lock (_dispatchers)
                            _dispatchers[context.Dispatcher] -= count;
                        break;
                    }

                }
                return context.CurrentResult;
            }
```

**Comment by rprouse** [comment]
It looks like you solved your problem, so I am going to close this as a question.
